feat(build_wheels): 1st step of rewritting the wheel builds (IDF-8438)

jakub-kocka opened the pull request:
Assembled the list of packages.
From all branches of IDF starting 5.0 and corresponding constraints files.

- Used `set` data type instead of the `list` because in `set` no duplicands are allowed

1. However the set is used, there can be not exact duplicands with another versions and so on. After implementing the `include-list` and `exclude-list` this will be formed into single requirement depending on the weight of files combined and logic

- Used `Requirement` from Python package `packaging` to store the requirements with additional markers (such as python version, platform, etc.)
- `include-list` is already there and it is file of requirements which are in IDF 4.3 and 4.4 version and not in the newer versions 

The Python script is executed in [workflow](https://github.com/espressif/idf-python-wheels/actions/runs/6716902139/job/18253837725?pr=20), under the `"Build wheels for IDF"` job is printed the assembled list of requirements.

dobairoland commented on .github/workflows/build-wheels-windows-dispatch_new.yml:
```diff
@@ -0,0 +1,33 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ['3.8.10']
```
Is this version motivated by the minimal ESP-IDF requirement or the Windows installer?

Windows installer used to deploy 3.8 for years. However, it has recently switched to 3.11.

Can you check please if the wheels (e.g. cryptography) built for 3.8 are usable for 3.11?

jakub-kocka replied:
This was motived by minimal ESP-IDF requirement.

There is only Python 3.6 version for Windows so I tested it with this for Python 3.11 and it was successfully installed.

dobairoland replied:
But now you've switched to 3.11. Why?

jakub-kocka replied:
Thank you for personal clarification, switched back.

dobairoland commented on Build-Wheels.py:
```diff
@@ -0,0 +1,66 @@
+
+# Global variables for branches and constraints files for the branches to be downloaded
+# (length of each list should be the same and so the positions for branch and it's specific constraints file)
+IDF_BRANCH = ['release/v5.0', 'release/v5.1', 'master']
```
Master is already v5.3, so v5.2 is missing.

Can't we use some kind of version detection? It will frequently happen that master becomes newer than the number of the second list indicates. 

Also in cases when the list is constant, it is better to use tuple which is non-mutable and makes clear to the reader that the list won't be changed.

jakub-kocka replied:
Thank you @dobairoland, this is true but there is not v5.3 release on `Github` yet.
So this brings me to think if this whole `idf-python-wheels` shouldn't have to be on `Gitlab` as well to already have build wheels when there is a `Github` release?


Actually I have been already thinking about this suggestion before you have made it and I have tried to make this fully automated now.

dobairoland replied:
It is better to handle these builds in one place instead of two. 

I think the main motivation behind having this here is the availability of Windows runners. For the self hosted runners I guess the internal repository would be better.

So IMO we should leave this project on Github, unless @tomassebestik thinks otherwise.

dobairoland commented on Build-Wheels.py:
```diff
@@ -0,0 +1,66 @@
+# Global variables for branches and constraints files for the branches to be downloaded
+# (length of each list should be the same and so the positions for branch and it's specific constraints file)
+IDF_BRANCH = ['release/v5.0', 'release/v5.1', 'master']
+IDF_CONSTRAINTS = ['v5.0', 'v5.1', 'v5.2']
```
Having two lists which dependent each other is very dangerous.

- The best would be to detect the IDF version - e.g. from the cmake file (see the implementation of `idf.py --version`). 
- Another solution would be to get it from the branch name. For master it would fall back to a variable with would define the version for master. It is still better than having two lists.

dobairoland commented on Build-Wheels.py:
```diff
@@ -0,0 +1,66 @@
+requirements_set = set()
+
+# Download all requirements from all the branches requirements and constraints files
+for i in range(len(IDF_BRANCH)):
```
This is the Pythonic and faster way to get the index while iterating through the list:
```python
for i, item in enumerate(IDF_BRANCH):
```

dobairoland commented on Build-Wheels.py:
```diff
@@ -0,0 +1,66 @@
+        print(f'Failed to download ESP-IDF constraints for branch {IDF_BRANCH[i]}')
+
+    # Add requirements into set to prevent duplicates
+    for line in requirements_txt:
```
You can do 
```python
for line in map(str.strip, requirements_txt):
```

and then you don't need to strip later and you can check `#` at the beginning of the line properly.

- I don't think `'#' not in line` is correct. E.g. comment can be at the end of the line.
- You might also don't match `\n` if it is a Windows EOL or contains other white spaces.

dobairoland commented on include-list.txt:
```diff
@@ -0,0 +1,15 @@
+# This is a list of python additional packages needed for older versions of ESP-IDF (v4.3, v4.4). This file is used with pip.
```
We might stress out that this list is not just for those cases. 

At some point we need to start to update the README as well. This file should be mentioned there. Or where to update the actively supported IDF versions. For this reasons, those variables might be placed in a separate file, e.g. `constants.py`.

This can be started now or left to the end.

dobairoland commented on Build-Wheels.py:
```diff
@@ -0,0 +1,66 @@
+    for feature in idf_requirements_json['features']:
+        r = requests.get(f"{idf_resources_url}{IDF_BRANCH[0]}/{feature['requirement_path']}")
+        if r.status_code == 200:
+            if not feature['optional'] or install_optional:
```
We want to build wheels for optional features as well.

dobairoland reviewed (changes requested):
Thanks @jakub-kocka. Nice work! I've left some suggestions.

peterdragun commented on Build-Wheels.py:
```diff
@@ -0,0 +1,66 @@
+            try:
+                requirements_set.add(Requirement(line.strip()))
+            except InvalidRequirement:
+                # TODO not a classic requirement
```
I am just curious, what is example of 'not a classic requirement'? 🤔 Maybe it would be nice to include a comment with example when you finish the implementation

jakub-kocka replied:
Yes, that is a good idea. I am not sure how this will be implemented yet.

The example of 'not a classic requirement' is for example: `--only-binary cryptography`

peterdragun commented on .github/workflows/build-wheels-windows-dispatch_new.yml:
```diff
@@ -0,0 +1,33 @@
+        run: pip install -r .\build_requirements.txt
+
+      - name: Build wheels for IDF
+        run: python .\Build-Wheels.py
```
Optional: I would personally change the name of `Build-Wheels.py` file to something more standard in python like `build_wheels.py`. I can see that you wanted to keep the name from the original ps script but this looks a bit strange to me.

jakub-kocka replied:
That is true, thank you.

fhrbata commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+
+# ESP-IDF branches list
+IDF_BRANCHES_URL = 'https://api.github.com/repos/espressif/esp-idf/branches'
+r = requests.get(IDF_BRANCHES_URL)
```
Can we limit this to protected branches only and bump the `per_page` limit if we don't consider possible paging?

`https://api.github.com/repos/espressif/esp-idf/branches?protected=true&per_page=100`

It's not problem now, because there is just 18 branches, but who knows if this doesn't change in the future.

jakub-kocka replied:
Thank you Frantisek for good suggestion, changed this.

dobairoland commented on .github/workflows/build-wheels-windows-dispatch_new.yml:
Please don't put "new" into any of the file names. When we remove old stuff we don't want to do refactoring as well and have even more changes in the PR.

dobairoland commented on include_list.txt:
```diff
@@ -0,0 +1,15 @@
+
+# ----- IDF v4.3 and v4.4 -----
+# cffi 1.15 supports Python 3.6+ only
+cffi<1.15;python_version<'3.6'
```
We won't build for old Pythons. So this line can be removed.

dobairoland commented on include_list.txt:
```diff
@@ -0,0 +1,15 @@
+cffi<1.15;python_version<'3.6'
+
+# esptool requirements (see components/esptool_py/esptool/setup.py)
+reedsolo>=1.5.3,<=1.5.4
```
If we add this requirement in addition to the latest esptool's `"reedsolo>=1.5.3,<1.8"` (https://github.com/espressif/esptool/blob/master/setup.py#L131C9-L131C31) then won't that mean that we will build only `reedsolo==1.5.4`?

If an ESP-IDF v5.x user will install esptool then it will pull in `reedsolo==1.7.0` and the wheel will be built. Or am I wrong?

IMO, you can remove this restriction as well. `reedsolo` is a pure-Python package for which we don't need to provide a wheel.

dobairoland commented on include_list.txt:
```diff
@@ -0,0 +1,15 @@
+# esptool requirements (see components/esptool_py/esptool/setup.py)
+reedsolo>=1.5.3,<=1.5.4
+# bitstring 4 dropped support for Python 3.6 and older
+bitstring>=3.1.6,<4
```
This is also pure-Python.

dobairoland commented on include_list.txt:
```diff
@@ -0,0 +1,15 @@
+reedsolo>=1.5.3,<=1.5.4
+# bitstring 4 dropped support for Python 3.6 and older
+bitstring>=3.1.6,<4
+ecdsa>=0.16.0
```
Also pure Python.

dobairoland commented on include_list.txt:
```diff
@@ -0,0 +1,15 @@
+ecdsa>=0.16.0
+
+# espcoredump requirements
+# This is the last version supports both 2.7 and 3.4
```
Also we don't build for these. `construct` is pure Python so I think we can remove the following restriction.

dobairoland commented on README.md:
```diff
@@ -39,3 +39,15 @@ The build contains all wheels required by branches:
+These lists are .txt files for requirements which should be added or excluded from the main requirements list which is automatically assembled.
+
+### include_list.txt
+File for additional Python packages to the main requirements list.
```
Just to be clear, this additions should be added in a separate build in order to not restrict the "main list". The restrictions are put there by the exclude list.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+AUTO_VERSION_CONSTRAINS_FILE = True
+
+# ESP-IDF branches list
+IDF_BRANCHES_URL = 'https://api.github.com/repos/espressif/esp-idf/branches?protected=true&per_page=100'
```
Oh, this is nice idea! Thanks!

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+idf_branches = []
+for branch in idf_repo_branches:
+    if branch.startswith('release'):
+        if float(branch.split('v')[1]) >= 5.0:
```
Float? You can compare tuples. `(A, B) >= (C, D)`

dobairoland replied:
If I might suggest, then just make a regex for `release/vX.Y` and use `(int(X), int(Y))` for comparison.

Also this 5.0 could be a constant which can be changed easily when we switch to 6.0.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+            else:
+                idf_constraints.append(idf_constraints[-1])
+        else:
+            idf_constraints.append('6.0')
```
What is this hard-coded 6.0?

jakub-kocka replied:
This is in case the `AUTO_VERSION_CONSTRAINS_FILE` will be `false`. I was thinking if the last IDF release version will be `v5.X` the auto version will try to assign for master branch `v5.(X+1)` but there will be the major release planned and master will be `v6.0`.

Yes, there is an fallback option for last release constraints file but I thought it could be needed to match master with the right constraints file.

But I should leave a comment in code for this, thanks.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+        splitted_branch = branch.split('release/')
+        idf_constraints.append(splitted_branch[1])
+    except IndexError:
+        if AUTO_VERSION_CONSTRAINS_FILE:
```
Why do we have a variable for this? Do you plan to overwrite it with and environment variable?

jakub-kocka replied:
Yes, as I thought about something like this as I mentioned above with "hard-coded 6.0".

And speaking of that for the `'6.0' ` in `idf_constraints.append('6.0')` could be a variable too.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+        idf_constraints.append(splitted_branch[1])
+    except IndexError:
+        if AUTO_VERSION_CONSTRAINS_FILE:
+            idf_last_release = idf_constraints[-1].split('v')[1].split('.')
```
Please just use regular expressions instead of all of these splittings.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+
+    # Add requirements into set to prevent duplicates
+    for line in map(str.strip, requirements_txt):
+        if line != '\n' or line != '\r\n' or line != '\r':
```
Are these checks meaningful? If we stripped the lines then I think none of these combinations are possible. On the other hand, you are missing the check for empty lines.

jakub-kocka replied:
You are right, thank you.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+    for line in map(str.strip, requirements_txt):
+        if line != '\n' or line != '\r\n' or line != '\r':
+            # check if in the line or the line itself is not a comment
+            if '#' in line:
```
1. This check will go through the whole line and will check each character. (for all lines)
2. If match has been found then it will go again essentially doing the same and splitting.

So you can do the following for every line
```python
line = line.split('#', 1)[0]
```
which at least saves no 2. and makes a code simpler by removing one "if" branch.

Please let me know if these kind of optimizations doesn't make sense to you. I can explain more.

dobairoland commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+
+
+##############################################################
+# TODO setuptools requirement should be first
```
I think is not necessary anymore. We started with support of fairly old Pythons (and systems). I don't think some OS would would shift Python 3.8+ with that old setuptools. Beside that, we use virtual environments only where setuptools is updated upon every environment creation (https://github.com/espressif/esp-idf/blob/master/tools/idf_tools.py#L2161).

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,113 @@
+            # check if in the line or the line itself is not a comment
+            if '#' in line:
+                splitted_line = line.split('#')
+                if splitted_line[0] != '':
```
Unnecessary `else` - this is perhaps better to write like (guard):
```py
                if splitted_line[0]== '':
                    continue
                line = splitted_line[0]
  
``` ?

tomassebestik commented on .github/workflows/build-wheels-windows-dispatch_new.yml:
```diff
@@ -0,0 +1,33 @@
+name: windows-dispatch
+
+on:
+  pull_request:
```
How is this definition supposed to work? If you set it up this way, it will be triggered on every push to any branch of any PR to this repository. 

Is that how it is intended?

jakub-kocka replied:
Thank you.
No it is not, it's just for now to trigger the workflow check.
I should leave a comment here or we can merge into master right version with schedule.

jakub-kocka commented:
Thank you for reviews and in person discussion @dobairoland, @fhrbata, @tomassebestik.

I have implemented your suggestions.

And as the code started to be not so transparent I have tried to do refactoring of it when it gets more complex to be still maintainable

- Code refactor into functions with docstrings
- Added constants.py for specifying some funcionalities
- [workflow run](https://github.com/espressif/idf-python-wheels/actions/runs/6733111466/job/18301099818?pr=20) under `Build wheels for IDF`

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,227 @@
+
+
+# Constraints files versions list
+def get_constraints_versions(idf_branches: list) -> list:
```
This nesting functions look suspicious 🤨  ....

Perhaps you could consider implement it with guard/short-circuit style like this:
```py
def get_constraints_versions(idf_branches: List[str]) -> List[str]:
    """From desired branches passed in get constraints files versions list."""
    idf_constraints = []

    for branch in idf_branches:
        # Handle release branches
        if 'release/' in branch:
            idf_constraints.append(branch.split('release/')[1])
            continue

        # Handle master branch
        if not constants.AUTO_VERSION_CONSTRAINS_FILE:
            idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
            continue

        # If there are no previous constraints, use the master constraints file
        if not idf_constraints:
            idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
            continue

        # Parse the last version number and increment the minor version
        match = re.match(r'v(\d+)\.(\d+)', idf_constraints[-1])
        if not match:
            # No valid version found in the last constraint, use the master constraints file
            idf_constraints.append(constants.MASTER_CONSTRAINS_FILE)
            continue  # Skip to the next iteration

        # A valid version was found, increment the minor version
        last_major, last_minor = map(int, match.groups())
        next_minor_version = f'v{last_major}.{last_minor + 1}'
        response = requests.get(f'{IDF_CONSTRAINTS_URL}{next_minor_version}.txt', timeout=10)
        # Append the next minor version if it exists, otherwise append the last known good constraints
        idf_constraints.append(next_minor_version if response.status_code == 200 else idf_constraints[-1])

    return idf_constraints
```

This is more flat and easy to read and maintain.

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,227 @@
+
+import constants
+
+idf_constraints_url = 'https://dl.espressif.com/dl/esp-idf/espidf.constraints.'
```
This is constant, so it should be upper-case:
```py
IDF_CONSTRAINTS_URL= 'https://dl.espressif.com/dl/esp-idf/espidf.constraints.'
```
.... and possibly in the file with other your constants (`constants.py`)

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,227 @@
+    idf_resources_url = 'https://raw.githubusercontent.com/espressif/esp-idf/'
+    requirements_txt: list = []
+
+    def download_branch_requirements() -> list:
```
I would prefer not nested functions.

For example, if `def download_branch_requirements() -> list:` is a helper function, you can mark it as such with an underscore before the function name and move it up a level:

```py
def _download_branch_requirements(branch: str, idf_requirements_json: dict) -> list:
    """Download requirements files for all groups specified in IDF requirements.JSON."""
    print(f'---------- ESP-IDF BRANCH {branch} ----------')
    requirements_txt = []
    for feature in idf_requirements_json['features']:
        res = requests.get(f"{IDF_RESOURCES_URL}{branch}/{feature['requirement_path']}")
        if check_response(res, f"Failed to download feature (requirement group) '{feature['name']}'"):
            requirements_txt.extend(res.text.splitlines())
            print(f"Added ESP-IDF {feature['name']} requirements")
    return requirements_txt
```

The same for `_download_branch_constraints` and `_add_into_requirements`.

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,227 @@
+        constraint_file_url = f'https://dl.espressif.com/dl/esp-idf/espidf.constraints.{idf_constraints[i]}.txt'
+
+        res = requests.get(idf_requirements_json_url)
+        if check_response(res, f'\nFailed to download requirements JSON for branch {branch}'):
```
Unecessary else:
```py
        if not check_response(res, f'\nFailed to download requirements JSON for branch {branch}'):
            continue

        idf_requirements_json = json.loads(res.content)
        requirements_txt.extend(_download_branch_requirements(branch, idf_requirements_json))
        requirements_txt.extend(_download_branch_constraints(branch, idf_constraints[i]))
```

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,227 @@
+##############################################################
+def main():
+    """Builds Python wheels for IDF dependencies
+    - Automatically fetch IDF branches and corresponding constraints files
```
These details in doc-string seem unnecessary, because the same is obvious from the names of the called functions (you have them well named).

tomassebestik reviewed (commented):
Thank you @jakub-kocka for the updates. I like the changes you made to better code structure. 

I left a few suggestion (feel free to ignore code style suggestions if your way seems to you better).

jakub-kocka commented on build_wheels.py:
```diff
@@ -0,0 +1,227 @@
+    Returns:
+        bool: True if status code 200
+    """
+    if response.status_code != 200:
```
You are right, thank for review.

jakub-kocka commented on constants.py:
```diff
@@ -0,0 +1,18 @@
+#
+# This feature can be overwritten by setting the AUTO_VERSION_CONSTRAINS_FILE to False
+# then the value in MASTER_CONSTRAIN_FILE is taken
+AUTO_VERSION_CONSTRAINS_FILE: bool = True
```
Thank you @tomassebestik, if I am not wrong `AUTO_VERSION_CONSTRAINS_FILE=false` will be needed if last `release` version of IDF will be `v5.X` and `master` will be planed to have major release to `v6.0`, the auto-version mechanism for constraints file will be assigning the `master` branch `v5.(X+1)` constraints file which will not exist and than it falls back to last release constraints file version for `master`.

So in that case I thought it might be useful to have "force overwrite" option of constraints file for `master` branch.

dobairoland replied:
But after releasing v6.1, one also has to go and set back AUTO_VERSION_CONSTRAINS_FILE to true.

I don't like the fact that v6.0 is hard-coded in two different files as well. We will also have to remember that those need to be set to v7.0 as well?

Instead of having such a complicated flow why couldn't we just download https://github.com/espressif/esp-idf/blob/master/tools/cmake/version.cmake (without cloning) and get the version from it? We wouldn't have to make any modification in the future.

tomassebestik commented on build_wheels.py:
```diff
@@ -0,0 +1,195 @@
+# This feature can be overwritten by setting the AUTO_VERSION_CONSTRAINS_FILE to False
+# in Github variables or manual workflow
+# then the value in MASTER_CONSTRAIN_FILE is taken
+AUTO_VERSION_CONSTRAINS_FILE: bool = bool(os.environ.get('AUTO_VERSION_CONSTRAINS_FILE', 'True'))
```
@jakub-kocka  I don't think it's going to work the way you probably think...

If the string `False` is passed as value of `AUTO_VERSION_CONSTRAINS_FILE`, it will be bool True (a non-empty string is True).

Check the lib `distutils.util -> strtobool` (up to Python 3.10) or `distlib.util -> strtobool` (python 3.10 and newer, this one needs to be installed `pip install distlib`)

Alternatively you can use 1/0 as switch or do something like this:

```py
AUTO_VERSION_CONSTRAINS_FILE: bool = os.environ.get('AUTO_VERSION_CONSTRAINS_FILE', 'True').lower() == 'true'
```

jakub-kocka replied:
Thank you, @tomassebestik. That is right.

jakub-kocka commented:
Can you please take a look @dobairoland, @fhrbata, @tomassebestik.

I have a final state for this first part of wheels build improvement. Thanks for your suggestions.

- Moved nested functions into global scope
- Implemented Github variables into workflow file and an option to rewrite these variables when running workflow manually

Unfortunately Github variables can not be checked when PR is from forked repo, so for the next steps I will make a branch in this repository.

The next steps will be to add a logic of building wheels from gathered requirements and implement an `include_list` and `exclude_list` into this logic.

fhrbata commented on build_wheels.py:
```diff
@@ -0,0 +1,199 @@
+            continue
+
+        # Handle master branch - NO automatic version mechanism
+        if not AUTO_VERSION_CONSTRAINS_FILE or idf_branches[0] == 'master':
```
Hi @jakub-kocka , I'm confused by this condition. Why is the `idf_branches[0] == 'master'` part there. We know that the branches could be only `release/vX.Y` or `master`. This is assured in `get_used_idf_branches`. Also I'm not sure how `master` can be first entry in the `idf_branches` list. Wouldn't just simple
```
if not AUTO_VERSION_CONSTRAINS_FILE
```
be enough here? We know that it has to be `master` at this point right?
Thank you

jakub-kocka replied:
Hi, @fhrbata. This was meant to be just a failsafe when the `MIN_IDF_MAJOR_VERSION` and `MIN_IDF_MINOR_VERSION` will be set in GitHub variables to not suit any release version of IDF (grater than).

If so and the `AUTO_VERSION_CONSTRAINS_FILE` will be **left to `True`**, there will be no possible way to apply auto-mechanism followed -> uncaught Exception.

So this is a fall-back to take at least variable defined in `MASTER_CONSTRAINS_FILE` GitHub variable because `master` will be in the `idf_branches`  always **(0 position if no` release`)**.

That means `MASTER_CONSTRAINS_FILE` should have a valid constraints file version ... for now it could be 'v5.3' (`release/v5.2` is not in GitHub so the `v5.2` constraints file is taken for `master` now).

However if the `MASTER_CONSTRAINS_FILE` will not be valid, such as `v6.0` the constrains file download will be skipped and requirements only for `master` branch without constraints file will be taken.


**So to make short conclusion, it is for fall-back options to not rise an uncaught Exception or to directly exit the script when bad inputs of GitHub variables will be provided and other requirements can be taken.**

This was the idea behind if it is bad or should be anything else I am open to this.

jakub-kocka replied:
Okay, sorry. The IDF `release/v5.2` is already on GitHub but the idea will be the same.

And thank you for review, I have discovered a mistake in `_download_branch_requirements` function thanks to this.

tomassebestik commented:
At this point I suggest to merge it and work on following parts of this task (I understood there is more steps).
Please don't forget to continue with branches instead of fork, that way you can test it properly in the repo.

LGTM, good job @jakub-kocka !

dobairoland commented on include_list.txt:
```diff
@@ -0,0 +1,2 @@
+# This is a list of Python additional packages needed. This file is used with pip.
+netifaces
```
If we add such item here then we need to comment why we did this. Otherwise, we don't know if or when can we removed if from the list. Is this not a dependency of pytest or ci packages?

I think Marek needed this just for one architecture.


This reminds me that I've proposed a different format for these files in IDF-8157. 

```
The format could be a YML file with records about the python version, architecture, comment and the package requirement (name and version) [also could be a list]. The python version and architecture could be "*" or a list of desired ones. Each build action then could check if it is in the list to build the package.
```
It is OK to do it later if you haven't forget about it.

jakub-kocka commented on .github/workflows/build-wheels-windows.yml:
```diff
@@ -0,0 +1,49 @@
+name: windows-dispatch
+
+# TODO schedule for workflow and delete pull_request/push
```
Not necessarily but it would be better to delete every automatic run until it is completed. So yes, I will delete it and leave only `on: workflow_dispatch` before merge for now.

jakub-kocka commented:
@dobairoland now it should be ready to merge.

- Adjusted function for getting master branch constraints file version, thanks for in person suggestion

also realised we actually don't need to fall-back with version of constraints file for master (we just don't download it and make requirements from other downloaded files)

- Also deleted `include_list` and `exclude_list` when logic was not clearly specified yet.
- Squashed commit for merge

jakub-kocka commented on .github/workflows/build-wheels-windows.yml:
```diff
@@ -0,0 +1,47 @@
+on:
+  workflow_dispatch:
+    inputs:
+      auto_version_constrains_file_in:
```
Thank you, I have forgot about this. Should be correct by now.
